test/Streamly/Test/Internal/Data/Fold.hs

```diff
@@ -1,27 +1,215 @@
+lesser f x y = if f x y == LT then x else y
+
+
+greater :: (a -> a -> Ordering) -> a -> a -> a
+greater f x y = if f x y == GT then x else y
+
+
+foldMaybe :: (b -> a -> b) -> b -> [a] -> Maybe b
+foldMaybe f acc ls =
+                  case ls of
+                     [] -> Nothing
+                     _ -> Just (foldl f acc ls)
+
+
+maximumBy :: (Ord a, Show a) => a -> (a -> a -> Ordering) -> [a] -> Expectation
+maximumBy genmin f ls = S.fold (FL.maximumBy f) (S.fromList ls) `shouldReturn` foldMaybe (greater f) genmin ls
+
+
+maximum :: (Show a, Ord a) => a -> [a] -> Expectation
+maximum genmin ls = S.fold FL.maximum (S.fromList ls) `shouldReturn` foldMaybe (greater compare) genmin ls
+
+
+minimumBy :: (Ord a, Show a) => a -> (a -> a -> Ordering) -> [a] -> Expectation
+minimumBy genmax f ls = S.fold (FL.minimumBy f) (S.fromList ls) `shouldReturn` foldMaybe (lesser f) genmax ls
+
+
+minimum :: (Show a, Ord a) => a -> [a] -> Expectation
+minimum genmax ls = S.fold FL.minimum (S.fromList ls) `shouldReturn` foldMaybe (lesser compare) genmax ls
+
+
+toList :: [Int] -> Expectation
+toList ls = S.fold FL.toList (S.fromList ls) `shouldReturn` ls
+
+
+safeLast :: [a] -> Maybe a
+safeLast [] = Nothing
+safeLast (x:[]) = Just x
+safeLast (_:xs) = safeLast xs
+
+
+last :: [String] -> Expectation
+last ls = S.fold FL.last (S.fromList ls) `shouldReturn` safeLast ls
+
+
+nth :: Int -> [a] -> Maybe a
+nth idx (x:xs) = if idx == 0
+                 then Just x
+                 else if idx < 0
+                 then Nothing
+                 else nth (idx - 1) xs
+nth _ [] = Nothing
+
+
+index :: Int -> [String] -> Expectation
+index idx ls = let x = S.fold (FL.index idx) (S.fromList ls)
+                in x `shouldReturn` (nth idx ls)
+
+
+find :: (Show a, Eq a) => (a -> Bool) -> [a] -> Expectation
+find f ls = do
+               let x = S.fold (FL.findIndex f) (S.fromList ls)
+               y <- x
+               case y of
+                  Nothing  -> S.fold (FL.find f) (S.fromList ls) `shouldReturn` Nothing
+                  Just idx -> S.fold (FL.any f) (S.fromList $ take idx ls) `shouldReturn` False
+
+
+neg :: (a -> Bool) -> a -> Bool
+neg f x = if f x == True then False else True
+
+findIndex :: (a -> Bool) -> [a] -> Expectation
+findIndex f ls = do
+                     let x = S.fold (FL.findIndex f) (S.fromList ls)
+                     y <- x
+                     case y of
+                        Nothing  -> S.fold (FL.all $ neg f) (S.fromList ls) `shouldReturn` True
+                        Just idx -> if idx == 0
+                                    then S.fold (FL.all f) (S.fromList []) `shouldReturn` True
+                                    else S.fold (FL.all f) (S.fromList $ (take idx ls)) `shouldReturn` False
+
+predicate :: Int -> Bool
+predicate x = if x * x < 100 then True else False
+
+elemIndex :: Int -> [Int] -> Expectation
+elemIndex elm ls = do
+                        let x = S.fold (FL.elemIndex elm) (S.fromList ls)
```

Comment:
I'm not a huge fan of this style of formatting.  Please reformat the do block.
You can use this for reference: [coding-style](https://github.com/composewell/haskell-dev/blob/master/coding-style.rst)

Reply:
In this link https://github.com/composewell/haskell-dev/blob/master/coding-style.rst#do-blocks
it's written that if the do is not being combined with the previous line, then it can be formatted as - 
```
do
    putStrLn "hello"
    putStrLn "hello"

```
That's what I tried to do manually.

Reply:
In this case `do` blocks can be combined with the previous line:
So instead of,
```
elemIndex :: Int -> [Int] -> Expectation
elemIndex elm ls = do
                                           let x = S.fold (FL.elemIndex elm) (S.fromList ls)
```
do,
```
elemIndex :: Int -> [Int] -> Expectation
elemIndex elm ls = do
    let x = S.fold (FL.elemIndex elm) (S.fromList ls)
```